### ISSUE TITLE
images built with docker v25 or later fail to push

### ISSUE THREAD

**Issue by kalensk** [concourse/registry-image-resource#356]
### Describe the bug

Images built with docker version 25 or later fail to push as shown in the below error.

For example, my task is pulling the latest docker in docker image (dind) to build my docker image. Using the latest dind tag which is docker version 26 fails, but using version 24-dind succeeds.
```
 - task: build_image
      privileged: true
      config:
        platform: linux
        image_resource:
          type: registry-image
          source:
            repository: docker
            tag: dind # tag: 24-dind succeeds
```

```
INFO[0000] pushing tag(s) dev                           
2024/03/28 01:42:37 existing blob: sha256:a7d483a98052b3c5108d6ed6d38caded5d2a787564ba0be1de08a5434f32a9db
2024/03/28 01:42:38 pushed blob: sha256:24c21885ea2bbe64a8a0b21901a61520772c747e98fb013d2cedf5bc23eecd4b
2024/03/28 01:42:39 pushed blob: sha256:617e845ae9918ce7686edff62e4c892c25ac172f75d75c76baef9abcbc8db052
2024/03/28 01:42:39 pushed blob: sha256:1249c494fb021137c410d273b1fbc3cfccf2daa4fb0c13ad223b31e052837235
2024/03/28 01:42:40 pushed blob: sha256:7ade3cec065838642366736d7d269cbefa2fea5eb839621bff0c520e00a18058
2024/03/28 01:42:41 pushed blob: sha256:052ad86d9dbd8a53393b7700d75dbd25764e564663bcc1a763341e3ac8b1c2af

ERRO[0006] pushing image failed: pushing tag(s): PUT https://gcr.io/v2/my-project/image/manifests/dev: multiple errors returned: BLOB_UNKNOWN: Manifest "sha256:b3322b3a5550897191dbfec1f4a7a256f54f8a6874a6de1f3a1bc3da96e1f4d1" blob "sha256:31a3635cf85a9285b9aebc8e97a2e4021fed41b53eecd55e614a89a03fa1daa4" unknown to registry.; BLOB_UNKNOWN: Manifest "sha256:b3322b3a5550897191dbfec1f4a7a256f54f8a6874a6de1f3a1bc3da96e1f4d1" blob "sha256:3951720cdd18c039b98fd52d43853e3caf4c365d2aaca115c7f3a24f6ea6c07a" unknown to registry.; BLOB_UNKNOWN: Manifest "sha256:b3322b3a5550897191dbfec1f4a7a256f54f8a6874a6de1f3a1bc3da96e1f4d1" blob "sha256:4ae1d13ade680629cf4fbf0c7be28bd2442ab0e670883b7112e59f2aa0f58077" unknown to registry.; BLOB_UNKNOWN: Manifest "sha256:b3322b3a5550897191dbfec1f4a7a256f54f8a6874a6de1f3a1bc3da96e1f4d1" blob "sha256:9cbd3f40c29cbdc8801087adff6e90a065aab4fc1517c962c480ef8b8bebc909" unknown to registry.; BLOB_UNKNOWN: Manifest "sha256:b3322b3a5550897191dbfec1f4a7a256f54f8a6874a6de1f3a1bc3da96e1f4d1" blob "sha256:abdf7e6fb3df9898c3fa4b0985e1e967131a4ab884eccefaa9c0f99885d21a64" unknown to registry. 
```

### Reproduction steps

1. Push an image built with the latest docker.

### Expected behavior

Pushing an image should succeed.

cc: @xtremerui @cdmessin @pvaramballypivot

**Comment by xtremerui** [comment]
I think this is related to the image engine used in the registry. Have you tried with a different registry like docker hub?

**Comment by kalensk** [comment]
> I think this is related to the image engine used in the registry. Have you tried with a different registry like docker hub?

Unfortunately, I don't think it will be feasible to try that as I am pushing to a private registry in gcr.

**Comment by meezaan** [comment]
It's the v2 that gets inserted in the URL when pushing. 

> ERRO[0006] pushing image failed: pushing tag(s): PUT https://gcr.io/v2/my-project/image/manifests/dev: multiple errors returned: BLOB_UNKNOWN: Manifest ...

the `v2` does not exist on ghcr.io, just on docker hub?

I have the same issue when pushing to harbor.

**Comment by talset** [comment]
same here from an image build with docker 27.3.1

**Comment by taylorsilva** [comment]
I'm too lazy to try and reproduce this right now, but after the next Concourse release comes out (or someone just tries using the next version of this resource, v1.11.0) can someone report in and let us know if it's working now?

I'm assuming the issue is the go-containerregistry library. We were REALLY behind by quite a few versions. Skimming the release notes of the library it looks like there were some docker-engine related changes made, so I'm hopeful just bumping that dependency will fix this issue.

**Comment by talset** [comment]
> I'm too lazy to try and reproduce this right now, but after the next Concourse release comes out (or someone just tries using the next version of this resource, v1.11.0) can someone report in and let us know if it's working now?
> 
> I'm assuming the issue is the go-containerregistry library. We were REALLY behind by quite a few versions. Skimming the release notes of the library it looks like there were some docker-engine related changes made, so I'm hopeful just bumping that dependency will fix this issue.

I tried to simply update the go-containerregistry and rebuild the resource but it didn't fixed the issue. It might be linked to https://github.com/google/go-containerregistry/issues/2034

Anyway I will try the v1.11.0 and see ;)

**Comment by talset** [comment]
`v1.10.1` fail but I tried from a fresh build `master / 02ef7a2a40c501b5fa0426d54c3f4d1404f0ae1b`  

```
docker build . --build-arg base_image=paketobuildpacks/run-jammy-base:latest -t talset/test:latest
```

It’s working on my end. Would you like me to try with a different `base_image` ?

**Comment by taylorsilva** [comment]
Nope, I think that's enough info to give us confidence here. Thanks for picking up the slack! ❤ 

As an aside; we'll be looking at changing the base images in the near future. We need to if we want an ARM Concourse.
